modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts

```diff
@@ -100,14 +99,38 @@ export async function prove(p: bigint, q: bigint): Promise<DeserializedPaillierB
   // This is calculating the inverse of the function y^4 mod N,
   // i.e.y ^ (1 / 4), where N = pq is a blum integer using HOC - Fact 2.160
   // from cacr.uwaterloo.ca / hac / about / chap2.pdf
-  // Prover generates y_i.
+  // Prover generates a_i, b_i, and y_i.
   const y = generateY(N, w);
   // Prover calculates z_i = y_i ^ d mod N
   const z = y.map((y_i) => modPow(y_i, d, N));
   // Prover calculates x_i = y_i ^ 1/4 mod N using [HOC - Fact 2.160]
   const e = ((l + BigInt(4)) / BigInt(8)) ** BigInt(2);
-  const x = y.map((y_i) => modPow(y_i, e, N));
-  return { w, x, z };
+  const ab = Array(m).fill(null);
+  const x = await Promise.all(
+    y.map(async (y_i, i) => {
+      // Select random a_i, b_i so that y_i' = (-1)^{a_i} * w^{b_i} * y is
+      // quadratic residue of N using [HOC - Fact 2.137].
+      let t;
+      while (true) {
+        t = y_i;
+        ab[i] = (await randBits(2))[0];
```

Comment:
Since there're only 4 cases, I wonder if it's sufficient to just enumerate them, just in case random generation gives duplicate values.

Reply:
Done